eclipse/sw360
# feat(MultiGroupsForUsers): Added support for user to be part of multiple secondary groups and assume different secondary roles

> Added support for user to be part of multiple secondary groups and assume different secondary roles
> * Which issue is this pull request belonging to and how is it solving it? (*#1163*)
> * Did you add or update any new dependencies that are required for your change? - No

Issue: closes #1163 

### How To Test?
- Test Admin -> Users Page - edit Secondary Groups
- Old Functionality should be tested without secondary groups for `View/Create/Update/Delete/Moderation Request/Clearing Request` for `Project, Component, Releases, Licenses` 
- Validated correct permissions are given based on secondary Roles and whether it covers all scenarios
- Basic way to test this PR is. First Preference goes to Primary Roles. Then check for Secondary roles.
  - For example 
     - User from `GROUP1` has Primary role as `USER` but has secondary Roles as `ADMIN in GROUP2`. then in `GROUP1` he acts as `normal user` but in `GROUP2` he acts as `ADMIN`.
     - Test `Create/Update/Delete/List Projects`, `Search Projects`, `(Moderation Request - View/Accept/Decline for Project, Component, Releases, Licenses )` , `Clearing Request` on below logic like -
       - Currently if a CR belongs to different group , it is not visible/editable to user from other group, 
         - Now secondary group can be added and user can view CRs from other groups
       - Project from user1@sw360.org - DEPARTMENT with visibility as Group and Moderator.
         - Should not be visible/editable to user2@sw360.org - DEPARTMENT, Should be visible/editable by admin@sw360.org
         - Edit secondary department with different roles and test functionality
	   - Similarly for Moderation Request.
	   
**Note -** Both UI and REST API needs to be tested

- **Read/Write Rules**-	   
  - Clearing Request-
    - Readable, if user in same group as Project BusinessUnit
	- Writable/Editable - if User has write access to Project via Primary Roles/Secondary Roles `AND` User is has Primary Role  or Secondary role in that Group `atleast CLEARING_EXPERT`
  - Moderation Request
    - Readable/Writable/Editable 
	  - Project - Users with Primary Roles or Secondary Roles in that department atleast `Clearing Admin` and if Project is not closed -> creator, ProjectResponsible and Moderators
	  - Release/Component - Users with Primary Roles or Secondary Roles in that department atleast `Clearing Admin` and if Project is not closed -> creator and Moderators
	  - License - Users with Primary Roles or Secondary Roles in that department atleast `Clearing Admin`
  - Project
    - Read - According to Visibility including Secondary Group
	- Others
	
	```
	@Override
    public boolean isActionAllowed(RequestedAction action) {
        ImmutableSet<UserGroup> clearingAdminRoles=ImmutableSet.of(UserGroup.CLEARING_ADMIN,
                UserGroup.CLEARING_EXPERT);
        ImmutableSet<UserGroup> adminRoles=ImmutableSet.of(UserGroup.ADMIN,
                UserGroup.SW360_ADMIN);
        if (action == RequestedAction.READ) {
            return isVisible(user).test(document);
        } else if (document.getClearingState() == ProjectClearingState.CLOSED) {
            switch (action) {
                case WRITE:
                case ATTACHMENTS:
                    return PermissionUtils.isUserAtLeast(ADMIN, user) || isUserOfOwnGroupHasRole(clearingAdminRoles, UserGroup.CLEARING_ADMIN) || isUserOfOwnGroupHasRole(adminRoles, UserGroup.ADMIN);
                case DELETE:
                case USERS:
                case CLEARING:
                case WRITE_ECC:
                    return PermissionUtils.isAdmin(user) || isUserOfOwnGroupHasRole(adminRoles, UserGroup.ADMIN);
                default:
                    throw new IllegalArgumentException("Unknown action: " + action);
            }
        } else {
            return getStandardPermissions(action);
        }
    }
	
	
	protected boolean getStandardPermissions(RequestedAction action) {
        ImmutableSet<UserGroup> clearingAdminRoles=ImmutableSet.of(UserGroup.CLEARING_ADMIN,
                UserGroup.CLEARING_EXPERT);
        ImmutableSet<UserGroup> adminRoles=ImmutableSet.of(UserGroup.ADMIN,
                UserGroup.SW360_ADMIN);
        switch (action) {
            case READ:
                return true;
            case WRITE:
            case ATTACHMENTS:
                return PermissionUtils.isUserAtLeast(ADMIN, user) || isContributor() || isUserOfOwnGroupHasRole(clearingAdminRoles, UserGroup.CLEARING_ADMIN) || isUserOfOwnGroupHasRole(adminRoles, UserGroup.ADMIN);
            case DELETE:
            case USERS:
            case CLEARING:
                return PermissionUtils.isAdmin(user) || isModerator() || isUserOfOwnGroupHasRole(adminRoles, UserGroup.ADMIN);
            case WRITE_ECC:
                return PermissionUtils.isAdmin(user) || isUserOfOwnGroupHasRole(adminRoles, UserGroup.ADMIN);
            default:
                throw new IllegalArgumentException("Unknown action: " + action);
        }
    }
    ```
> Have you implemented any additional tests? - No

### Checklist
Must:

Signed-off-by: Jaideep Palit <jaideep.palit@siemens.com>
